pkg/bootstrap/platform/gcp.go

```diff
@@ -148,26 +158,39 @@ func (e *gcpEnv) Metadata() map[string]string {
 	if e.metadata != nil {
 		return e.metadata
 	}
-	envPid, envNPid, envCN, envLoc := parseGCPMetadata()
-	if envPid != "" {
+	envPid, envNPid, envCN, envLoc, err := parseGCPMetadata()
+
+	// If no error is returned, then we have a valid GCP environment.
+	// Else we query the API to discover the needed information
+	if err == nil {
 		md[GCPProject] = envPid
-	} else if pid, err := projectIDFn(); err == nil {
-		md[GCPProject] = pid
-	}
-	if envNPid != "" {
 		md[GCPProjectNumber] = envNPid
-	} else if npid, err := numericProjectIDFn(); err == nil {
-		md[GCPProjectNumber] = npid
-	}
-	if envLoc != "" {
 		md[GCPLocation] = envLoc
-	} else if l, err := clusterLocationFn(); err == nil {
-		md[GCPLocation] = l
-	}
-	if envCN != "" {
 		md[GCPCluster] = envCN
-	} else if cn, err := clusterNameFn(); err == nil {
-		md[GCPCluster] = cn
+	} else if IsMetadataEndpointAccessible() {
+		wg := sync.WaitGroup{}
+
+		suppliers := []metadataSupplier{
```

Comment:
@zirain if you have any suggestion/pattern that I can implement here to make it more readable, let me know and I will update the code accordingly

Reply:
I prefer to change the implements of `IsGCP`, just detected metadata endpoint?
@howardjohn WDYT?